services/sts/src/main/java/software/amazon/awssdk/services/sts/internal/StsWebIdentityCredentialsProviderFactory.java

```diff
@@ -0,0 +1,132 @@
+                                      .build();
+
+            AssumeRoleWithWebIdentityRequest request = AssumeRoleWithWebIdentityRequest.builder()
+                                                                                       .roleArn(roleArn)
+                                                                                       .roleSessionName(sessionName)
+                                                                                       .build();
+
+            AssumeRoleWithWebIdentityRequestSupplier supplier = new AssumeRoleWithWebIdentityRequestSupplier(request,
+                                                                                                             tokenFilePath);
+
+            this.credentialsProvider =
+                StsAssumeRoleWithWebIdentityCredentialsProvider.builder()
+                                                               .stsClient(stsClient)
+                                                               .refreshRequest(supplier)
+                                                               .build();
+        }
+
+        @Override
+        public AwsCredentials resolveCredentials() {
+            return this.credentialsProvider.resolveCredentials();
+        }
+
+        @Override
+        public void close() {
+            IoUtils.closeQuietly(credentialsProvider, null);
+            IoUtils.closeQuietly(stsClient, null);
+        }
+    }
+
+    private static final class AssumeRoleWithWebIdentityRequestSupplier implements Supplier {
+
+        private final AssumeRoleWithWebIdentityRequest request;
+        private final String webIdentityTokenFilePath;
+
+        AssumeRoleWithWebIdentityRequestSupplier(AssumeRoleWithWebIdentityRequest request,
+                                                        String webIdentityTokenFilePath) {
+            this.request = request;
+            this.webIdentityTokenFilePath = webIdentityTokenFilePath;
+        }
+
+        @Override
+        public Object get() {
+            return request.toBuilder().webIdentityToken(getToken(webIdentityTokenFilePath)).build();
+        }
+
+        private String getToken(String filePath) {
+            InputStream webIdentityTokenStream = FunctionalUtils.invokeSafely(() -> Files.newInputStream(Paths.get(filePath)));
+            String webIdentityToken = FunctionalUtils.invokeSafely(() -> IoUtils.toUtf8String(webIdentityTokenStream));
+
+            try {
+                return webIdentityToken;
+            } finally {
+                FunctionalUtils.invokeSafely(() -> IoUtils.closeQuietly(webIdentityTokenStream, null));
+            }
+        }
+    }
+
+
+    private static final class StsRetryCondition implements RetryCondition {
+
+        @Override
+        public boolean shouldRetry(RetryPolicyContext context) {
+            if (context.exception() instanceof IdpCommunicationErrorException) {
```

Comment:
nit: can be shortened to `return context.exception() instanceof IdpCommunicationErrorException`